### ISSUE TITLE
Comparison of speeds between VSSBlock and TransformerBlock

### ISSUE THREAD

**Issue by Jade999** [MzeroMiko/VMamba#45]
Hi,
Thank you for your excellent work! I have some questions when testing the inference efficiency of TransformerBlock and VSSBlock on a V100 gpu. 
The input size is set to the same that [16,16,128,128] of (n,c,h,w) format for TransformerBlock, and [16,128,128,16] of (n,h,w,c) format for VSSBlock. According to my understanding, the forward speed of VSSBlock should be much faster than that of Transformer, because the latter has the complexity of O (N^2), while the former has O (N). But the test results are the opposite，Did I make a mistake somewhere, or did you test and compare these two modules separately?

speed compare:
TranformerBlock: 0.0025s (avarage)
VSSBlock: 0.014s (avarage)

**Comment by MzeroMiko** [comment]
Can you share your benchmark with me? I tested the code with shape (b=16, c=16, h=128, w=128), and found that `vssblock` needs 7ms on `A100 + py310+cu121+torch2.2` while `torchvision.model.vision_transformer.EncoderBlock` needs 51ms. With backward, `vssblock` costs 66ms, while transformer encoder block costs 174ms.

Note that `flash-attention-v2` is integrated in [torch2.2](https://pytorch.org/blog/pytorch2-2/), which is the fastest implementation of attention we've ever known.



```python
def check_vssblock():
    import triton
    from torchvision.models.vision_transformer import EncoderBlock

    vb = VSSBlock(
        hidden_dim=16, 
        drop_path=0.0, 
        norm_layer=nn.LayerNorm, 
        ssm_d_state=1, 
        ssm_ratio=2, 
        ssm_rank_ratio=-1, 
        ssm_dt_rank="auto", 
        ssm_act_layer=nn.SiLU,
        ssm_conv=3, 
        ssm_conv_bias=False, 
        ssm_drop_rate=0.0, 
        ssm_simple_init=True, 
        forward_type="v2", 
        mlp_ratio=4, 
        mlp_act_layer=nn.GELU, 
        mlp_drop_rate=0.0, 
        use_checkpoint=False,
    ).cuda()
    
    trans = EncoderBlock(
        num_heads=1, 
        hidden_dim=16, 
        mlp_dim=int(4.0 * 16), 
        dropout=0.0, 
        attention_dropout=0.0, 
        norm_layer=nn.LayerNorm,
    ).cuda()

    inp = torch.randn((16, 128, 128, 16)).cuda().requires_grad_()
    inp2 = inp.detach().cuda().view(16, -1, 16).requires_grad_()
    fn = lambda :vb(inp)
    ms = triton.testing.do_bench(fn, warmup=100)
    print(ms)
    fn = lambda :trans(inp2)
    ms = triton.testing.do_bench(fn, warmup=100)
    print(ms)
    fn = lambda :vb(inp).sum().backward()
    ms = triton.testing.do_bench(fn, warmup=100)
    print(ms)
    fn = lambda :trans(inp2).sum().backward()
    ms = triton.testing.do_bench(fn, warmup=100)
    print(ms)
    import time; time.sleep(10000)
    # 7.127295970916748
    # 51.02694320678711
    # 65.73977661132812
    # 174.20799255371094
```

**Comment by Jade999** [comment]
Perhaps it's because what I used to compare is the Lite Transformer block。
`
##########################################################################
# Multi-DConv Head Transposed Self-Attention (MDTA)
class Attention(nn.Module):
    def __init__(self, dim, num_heads, bias):
        super(Attention, self).__init__()
        self.num_heads = num_heads
        self.temperature = nn.Parameter(torch.ones(num_heads, 1, 1))

        self.qkv = nn.Conv2d(dim, dim*3, kernel_size=1, bias=bias)
        self.qkv_dwconv = nn.Conv2d(
            dim*3, dim*3, kernel_size=3, stride=1, padding=1, groups=dim*3, bias=bias)
        self.project_out = nn.Conv2d(dim, dim, kernel_size=1, bias=bias)

    def forward(self, x):
        b, c, h, w = x.shape

        qkv = self.qkv_dwconv(self.qkv(x))
        q, k, v = qkv.chunk(3, dim=1)

        q = rearrange(q, 'b (head c) h w -> b head c (h w)',
                      head=self.num_heads)
        k = rearrange(k, 'b (head c) h w -> b head c (h w)',
                      head=self.num_heads)
        v = rearrange(v, 'b (head c) h w -> b head c (h w)',
                      head=self.num_heads)

        q = torch.nn.functional.normalize(q, dim=-1)
        k = torch.nn.functional.normalize(k, dim=-1)

        attn = (q @ k.transpose(-2, -1)) * self.temperature
        attn = attn.softmax(dim=-1)

        out = (attn @ v)

        out = rearrange(out, 'b head c (h w) -> b (head c) h w',
                        head=self.num_heads, h=h, w=w)

        out = self.project_out(out)
        return out

class LayerNorm(nn.Module):
    def __init__(self, dim, LayerNorm_type):
        super(LayerNorm, self).__init__()
        if LayerNorm_type == 'BiasFree':
            self.body = BiasFree_LayerNorm(dim)
        else:
            self.body = WithBias_LayerNorm(dim)

    def forward(self, x):
        h, w = x.shape[-2:]
        return to_4d(self.body(to_3d(x)), h, w)

class TransformerBlock(nn.Module):
    def __init__(self, dim, num_heads, ffn_expand_factor, bias, LayerNorm_type):
        super(TransformerBlock, self).__init__()

        self.norm1 = LayerNorm(dim, LayerNorm_type)
        self.attn = Attention(dim, num_heads, bias)
        self.norm2 = LayerNorm(dim, LayerNorm_type)
        self.ffn = FeedForward(dim, ffn_expand_factor, bias)

    def forward(self, x):
        x = x + self.attn(self.norm1(x))
        x = x + self.ffn(self.norm2(x))
        return x`
